tket/src/Transformations/Decomposition.cpp

```diff
@@ -419,15 +422,327 @@ Transform decompose_MolmerSorensen() {
+          break;
+        }
+        case 3: {
+          sub.append(
+              CircPool::TK2_using_ZZPhase(angles[0], angles[1], angles[2]));
+          break;
+        }
+        default:
+          throw NotValid("Number of ZZPhase invalid in decompose_TK2");
+      }
+      break;
+    }
+    default:
+      throw NotValid("Unrecognised target OpType in decompose_TK2");
+  }
+  return sub;
+}
+
+Transform decompose_TK2() { return decompose_TK2({}); }
+Transform decompose_TK2(const TwoQbFidelities &fid) {
+  if (fid.ZZMax_fidelity) {
+    if (*fid.ZZMax_fidelity < 0 || *fid.ZZMax_fidelity > 1) {
+      throw NotValid("ZZMax fidelity must be between 0 and 1.");
+    }
+  }
+  if (fid.CX_fidelity) {
+    if (*fid.CX_fidelity < 0 || *fid.CX_fidelity > 1) {
+      throw NotValid("CX fidelity must be between 0 and 1.");
+    }
+  }
+  if (fid.ZZMax_fidelity && fid.ZZPhase_fidelity) {
+    if (*fid.ZZMax_fidelity < (*fid.ZZPhase_fidelity)(.5)) {
+      throw NotValid(
+          "The ZZMax fidelity cannot be smaller than the ZZPhase(0.5) "
+          "fidelity");
+    }
+  }
+  return Transform([fid](Circuit &circ) {
+    bool success = false;
+
+    // The gates that we currently support.
+    const std::set fixed_angle({OpType::CX, OpType::ZZMax});
+    const std::set variable_angle({OpType::ZZPhase});
+
+    VertexList bin;
+    BGL_FORALL_VERTICES(v, circ.dag, DAG) {
+      if (circ.get_OpType_from_Vertex(v) != OpType::TK2) continue;
+
+      success = true;
+      auto params = circ.get_Op_ptr_from_Vertex(v)->get_params();
+      TKET_ASSERT(params.size() == 3);
+      std::array<Expr, 3> angles{params[0], params[1], params[2]};
+
+      Circuit sub = TK2_replacement(angles, fid);
+      bin.push_back(v);
+      circ.substitute(sub, v, Circuit::VertexDeletion::No);
+    }
+    circ.remove_vertices(
+        bin, Circuit::GraphRewiring::No, Circuit::VertexDeletion::Yes);
+
+    return success;
+  });
+}
+
 Transform decompose_ZZPhase() {
   return Transform([](Circuit &circ) {
     bool success = decompose_PhaseGadgets().apply(circ);
+    VertexList bin;
     BGL_FORALL_VERTICES(v, circ.dag, DAG) {
       if (circ.get_OpType_from_Vertex(v) == OpType::PhaseGadget) {
         const Op_ptr g = circ.get_Op_ptr_from_Vertex(v);
         circ.dag[v] = {get_op_ptr(OpType::ZZPhase, g->get_params()[0])};
+      } else if (circ.get_OpType_from_Vertex(v) == OpType::XXPhase) {
+        success = true;
+        const Op_ptr g = circ.get_Op_ptr_from_Vertex(v);
+        Circuit sub = CircPool::XXPhase_using_ZZPhase(g->get_params()[0]);
+        circ.substitute(sub, v, Circuit::VertexDeletion::No);
+        bin.push_back(v);
+      } else if (circ.get_OpType_from_Vertex(v) == OpType::YYPhase) {
+        success = true;
+        const Op_ptr g = circ.get_Op_ptr_from_Vertex(v);
+        Circuit sub = CircPool::YYPhase_using_ZZPhase(g->get_params()[0]);
+        circ.substitute(sub, v, Circuit::VertexDeletion::No);
+        bin.push_back(v);
       }
```

Comment:
I think it should be a `switch` statement on `circ.get_OpType_from_Vertex(v)` instead of `if` clauses.

(_This would also solve the problem of `circ.get_OpType_from_Vertex(v)` being recalculated multiple times, which is trivial, but unnecessary_).

Reply:
Done!